packages/apollo-server-defer/README.md

```diff
@@ -0,0 +1,173 @@
+
+#### Use Cases
+
+```graphql
+type Video {
+  # public data that can be cached
+  id: ID!
+  title: String
+  description: String
+  # progress is private to the user, and significantly more
+  # expensive to retrieve
+  progress: Progress
+  comments: [Comment]
+}
+
+type Progress {
+  percent: Float
+  position: Float
+}
+
+type Comment {
+  userId: ID!
+  text: String
+}
+
+# Example Query using @defer
+query {
+  video(id: $id) {
+    title: String
+    progress @defer {
+      percent
+      position
+    }
+    comments @defer {
+      userId
+      text
+    }
+  }
+}
+```
+
+Instead of holding the GraphQL response and page rendering until the entire query is resolved, @defer tells Apollo Server to return a partial query response if the deferred fields are not ready yet. In the example above, we see how it may be used to remove `progress` and `comments` from the intial query response. Apollo Server would then take care of resolving the rest of the deferred fields in the background, and stream them to the client when they are ready.
+
+```
+// Initial Response
+{
+  "asset": {
+    "title": "Westworld",
+  }
+}
+
+// Stream in deferred fields as patches
+{
+  "path": ["video", "progress"],
+  "data": {
+    "percent": 20,
+    "position": 20
+  }
+}
+{
+  "path": ["video", "comments"],
+  "data": [{
+    "userId": "123",
+    "text": "..."
+  }]
+}
+```
+
+#### Who needs this the most?
+
+- Media companies, banks, other data-heavy enterprise applications.
+
+#### Why is @defer an ideal solution?
+
+- Query remains super easy to read 👍
+- Stays true to Apollo's declarative approach to data loading - everything is encapsulated in a single query component 😎
+- Greatly reduces boilerplate code to manage multiple requests, a huge value-add for users that use both Apollo Server AND Client together.
+
+### Implementation Details
+
+#### Patch Data Format:
+
+- Proposed format for `ExecutionPatchResult`
+  ```graphql
+  {
+    path: [string]!
+    data: {}
+    errors: []
+  }
+  ```
+- Current transport assumes ExecutionResult so some changes need to be made.
+- We should use the same patch format for @defer/stream/live support.
+
+#### Apollo Server:
+
+In order to support @defer, there are significant changes to be made to the execution phase if GraphQL.
+
+- Roll our own derivative of graphql.js execution that supports deferred responses and streaming patches using observables.
+- Try to maximize code reuse by exporting types and utility functions from graphql.js, making a PR if necessary.
+- Restrict peer dependency on graphql.js to versions that we have tested with.
+- @skip and @include should take priority over @defer
+
+#### Apollo Client:
+
+- No changes required
+- Initial implementation of @defer support should come as a Apollo Link. Reads from a socket connection or some other event stream, keeps the partial response in memory, merging patches as they come and pushing it through the link stack.
```

Comment:
I don't think we should do it this way. `@defer` will need special traits for reading / bypassing the cache so I think we should build it into Apollo Client core. Its also important to make this a seamless experience for people and adding another link is not ideal